File: common/src/main/java/cc/turtl/chiselmon/feature/pc/tooltip/TooltipEntry.java
package cc.turtl.chiselmon.feature.pc.tooltip;

import cc.turtl.chiselmon.client.config.category.PCConfig;
import com.cobblemon.mod.common.pokemon.Pokemon;
import net.minecraft.network.chat.Component;

import java.util.function.Function;
import java.util.function.Predicate;

import static cc.turtl.chiselmon.core.util.format.ComponentUtilsKt.labelled;

public record TooltipEntry(
        String translationKey,
        Predicate<PCConfig.TooltipConfig> configCheck,
        Predicate<Pokemon> pokemonCheck,
        Function<Pokemon, Object> componentProvider
) {
    public boolean shouldDisplay(PCConfig.TooltipConfig config, Pokemon pokemon) {
        return configCheck.test(config) && pokemonCheck.test(pokemon);
    }

    public Component getComponent(Pokemon pokemon) {
        Object value = componentProvider.apply(pokemon);
        return labelled(
                Component.translatable("chiselmon.ui.label." + translationKey),
                value);
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/feature/pc/tooltip/TooltipEntry.kt
package cc.turtl.chiselmon.client.feature.pc.tooltip

import cc.turtl.chiselmon.client.config.category.PCConfig
import cc.turtl.chiselmon.core.util.format.labelled
import com.cobblemon.mod.common.pokemon.Pokemon
import net.minecraft.network.chat.Component

data class TooltipEntry(
    val translationKey: String,
    val configCheck: (PCConfig.TooltipConfig) -> Boolean,
    val pokemonCheck: (Pokemon) -> Boolean,
    val componentProvider: (Pokemon) -> Any
) {
    fun shouldDisplay(config: PCConfig.TooltipConfig, pokemon: Pokemon) =
        configCheck(config) && pokemonCheck(pokemon)

    fun getComponent(pokemon: Pokemon): Component =
        labelled(
            Component.translatable("chiselmon.ui.label.$translationKey"),
            componentProvider(pokemon)
        )
}